packages/@lwc/engine-core/src/framework/profiler.ts

```diff
@@ -19,12 +30,77 @@ export enum OperationId {
     renderedCallback = 4,
     disconnectedCallback = 5,
     errorCallback = 6,
+    globalHydrate = 7,
+    globalRehydrate = 8,
 }
 
 enum Phase {
     Start = 0,
     Stop = 1,
 }
+// Even if all the browser the engine supports implements the UserTiming API, we need to guard the measure APIs.
+// JSDom (used in Jest) for example doesn't implement the UserTiming APIs.
+const isUserTimingSupported: boolean =
+    typeof performance !== 'undefined' &&
+    typeof performance.mark === 'function' &&
+    typeof performance.clearMarks === 'function' &&
+    typeof performance.measure === 'function' &&
+    typeof performance.clearMeasures === 'function';
+
+function getMarkName(phase: string, vm: VM): string {
+    // Adding the VM idx to the mark name creates a unique mark name component instance. This is necessary to produce
+    // the right measures for components that are recursive.
+    return `${getComponentTag(vm)} - ${phase} - ${vm.idx}`;
+}
+
+function getMeasureName(phase: string, vm: VM): string {
+    return `${getComponentTag(vm)} - ${phase}`;
+}
+
+function start(markName: string) {
+    performance.mark(markName);
+}
+
+function end(measureName: string, markName: string) {
+    performance.measure(measureName, markName);
+
+    // Clear the created marks and measure to avoid filling the performance entries buffer.
+    // Note: Even if the entries get deleted, existing PerformanceObservers preserve a copy of those entries.
+    performance.clearMarks(markName);
+    performance.clearMarks(measureName);
```

Comment:
should this be `clearMeasures` instead?

Reply:
Good catch. Looks like this should actually be `clearMeasures`. But it was already wrong in the codebase: 

https://github.com/salesforce/lwc/blob/d93e2963d8d32b138ce9ed130d98e7bef84b3d74/packages/%40lwc/engine-core/src/framework/performance-timing.ts#L54-L55

Maybe I should handle it in a separate PR so it isn't blocked on this one?

Reply:
https://github.com/salesforce/lwc/pull/2421